Extensions/Content/Dnn.PersonaBar.Extensions/WebApps/Pages.Web/src/components/dnn-persona-bar-page-treeview/src/PersonaBarPageTreeviewInteractor.jsx

```diff
@@ -655,15 +657,45 @@ class PersonaBarPageTreeviewInteractor extends Component {
             isTreeviewExpanded
         } = this.state;
 
-        this.props._traverse((item, list, updateStore) => {
+        const traverseCallback = () => this.props._traverse((item, list, updateStore) => {
             if (item.hasOwnProperty("childListItems") && item.childListItems.length > 0) {
                 item.isOpen = (isTreeviewExpanded) ? false : true;
                 updateStore(list);
                 this.setState({
                     isTreeviewExpanded: !this.state.isTreeviewExpanded
                 });
+            } else if (item.childCount === 0) {
+                delete item.childListItems;
+                delete item.isOpen;
+                delete item.hasChildren;
+                updateStore(list);
+            }
+        });
+
+        const waitForChildList = (timeoutTime, callback) => new Promise((resolve, reject) => {
+            const check = () => {                
+                if(this.state.isChildLoaded) {
+                    callback();
+                    resolve();
+                } else if((timeoutTime -= 100) < 0) {
+                    reject();
+                } else {
+                    setTimeout(check, 100);
+                }                
             }
+
+            setTimeout(check, 100)
         });
+
+        if (this.props.multiplePagesAdded) {
+            this.props.addPagesComplete();
+            this.setState({ isChildLoaded: false });
+            this.loadAllChildList();
+
+            waitForChildList(2000, traverseCallback);
```

Comment:
There is no need for extra polling, why not return the promise from "loadAllChildList" (return Promise.all...) so you can use it like this: `this.loadAllChildList().then(traverseCallback);`

And if you want to apply a time limit to before calling the "traverseCallback" action then you can do it like below too:

```
const loadChildListStartTime = new Date();
this.loadAllChildList().then(function () {
    if (new Date() - loadChildListStartTime < 2000) {
        traverseCallback();
    }
});
```

Reply:
There are many cascaded promises and asynchronous calls going down for `loadAllChildList`. In this case, we are waiting until the `isChildLoaded` state is set to `true`. Without this callback approach, there is a chance to miss this state transition. Hence, we are checking every 100 ms if this state is set to true or not up to to 2 seconds. Unfortunately the `promise.then` approach fails to deliver what is required in this specific case, otherwise, you are right.